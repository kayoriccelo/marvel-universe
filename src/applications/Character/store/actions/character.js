import api from '../../../../services/api'
import * as Types from '../types'
import history from '../../../../history'


export const getListingCharactersAPI = params => {
    let search = params['search'] ? `?nameStartsWith=${params['search']}` : ''

    return dispatch => {
        return api.get(`/v1/public/characters${search}`).then(response => {
            dispatch({ type: Types.LISTING_CHARACTERS, payload: response.data.data.results })

            return response.data.results
        })
    }
}

export const getListingCharacters = _ => {
    return dispatch => dispatch({ type: Types.LISTING_CHARACTERS })
}

export const getListingSeries = character => {
    return dispatch => {
        return api.get(`/v1/public/characters/${character}/series`).then(response => {
            let series = response.data.data ? response.data.data.results : response.data

            dispatch({ type: Types.LISTING_SERIES, payload: series })

            return series
        })
    }
}

export const clearSeries = _ => {
    return dispatch => {
        dispatch({ type: Types.LISTING_SERIES, payload: [] })
    }
}

export const loadCharacter = params => {
    return dispatch => {
        return api.get(`/v1/public/characters/${params['id']}`).then(response => {
            if (response.data.data !== undefined) {
                let character = response.data.data.results[0]
                dispatch([{ type: Types.LOAD_CHARACTER, payload: character }, getListingSeries(character.id)])
            } else {
                let character = response.data
                dispatch({ type: Types.LOAD_CHARACTER, payload: character })
            }
        })
    }
}

export const updateCharacter = params => {
    return dispatch => {
        let itens = params['itens'].map(item => item.id === params['instance'].id ? params['instance'] : item)

        dispatch([
            { type: Types.UPDATE_CHARACTER, payload: params['instance'] },
            { type: Types.LISTING_CHARACTERS, payload: itens }
        ])

        history.push('/characters/listing')
    }
}