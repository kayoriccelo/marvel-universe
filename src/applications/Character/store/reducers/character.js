import * as Types from '../types'


export const initialState = {
    itens: [],
    series: [],
    instance: null
}

export const characterReducer = (state = initialState, action) => {
    switch (action.type) {

        case Types.LISTING_CHARACTERS:
            return { ...state, itens: action.payload || state.itens }

        case Types.LOAD_CHARACTER:
            return { ...state, instance: action.payload }

        case Types.UPDATE_CHARACTER:
            return { ...state, instance: action.payload }

        case Types.LISTING_SERIES:
            return { ...state, series: action.payload }

        default:
            return state
    }
}
