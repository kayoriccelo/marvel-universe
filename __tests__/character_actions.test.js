import configureMockStore from 'redux-mock-store'
import { initialState } from './character.config'

const listingSuccess = (itens) => ({ type: 'LISTING_CHARACTERS', payload: itens })
const updateSuccess = (character) => ({ type: 'UPDATE_CHARACTER', payload: character })

const mockStore = configureMockStore()
const store = mockStore({})

const inicialStateCharacter = {
    ...initialState,
    series: [],
    instance: {}
}

describe('Testing character actions', () => {
    it('Testing of action LISTING_CHARACTERS', () => {
        store.dispatch(listingSuccess(inicialStateCharacter.itens))

        expect(store.getActions()).toMatchSnapshot()

        expect(store.getActions()[0]['payload'].length === 3).toBe(true)
    })

    it('Testing of action UPDATE_CHARACTER', () => {

        store.dispatch(updateSuccess({
            "id": 1011334,
            "name": "Testing Update",
            "description": "",
            "modified": "2014-04-29T14:18:17-0400"
        }))

        expect(store.getActions()[1]['payload'] != undefined).toBe(true)

        expect(store.getActions()).toMatchSnapshot()
    })
})