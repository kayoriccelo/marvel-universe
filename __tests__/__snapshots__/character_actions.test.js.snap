// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing character actions Testing of action LISTING_CHARACTERS 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "description": "",
        "id": 1011334,
        "modified": "2014-04-29T14:18:17-0400",
        "name": "3-D Man",
        "resourceURI": "http://gateway.marvel.com/v1/public/characters/1011334",
        "series": Object {
          "available": 3,
          "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/series",
          "items": Array [
            Object {
              "name": "Avengers: The Initiative (2007 - 2010)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1945",
            },
            Object {
              "name": "Deadpool (1997 - 2002)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/2005",
            },
            Object {
              "name": "Marvel Premiere (1972 - 1981)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/2045",
            },
          ],
          "returned": 3,
        },
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
        },
      },
      Object {
        "description": "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
        "id": 1017100,
        "modified": "2013-09-18T15:54:04-0400",
        "name": "A-Bomb (HAS)",
        "resourceURI": "http://gateway.marvel.com/v1/public/characters/1017100",
        "series": Object {
          "available": 1,
          "collectionURI": "http://gateway.marvel.com/v1/public/characters/1017100/series",
          "items": Array [
            Object {
              "name": "Hulk (2008 - 2012)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/3374",
            },
          ],
          "returned": 1,
        },
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16",
        },
      },
      Object {
        "description": "AIM is a terrorist organization bent on destroying the world.",
        "id": 1009144,
        "modified": "2013-10-17T14:41:30-0400",
        "name": "A.I.M.",
        "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009144",
        "series": Object {
          "available": 33,
          "collectionURI": "http://gateway.marvel.com/v1/public/characters/1009144/series",
          "items": Array [
            Object {
              "name": "Ant-Man & the Wasp (2010 - 2011)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/13082",
            },
            Object {
              "name": "Avengers (1998 - 2004)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/354",
            },
            Object {
              "name": "Avengers (1963 - 1996)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1991",
            },
            Object {
              "name": "Avengers and Power Pack (2017)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23123",
            },
            Object {
              "name": "Avengers and Power Pack Assemble! (2006)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1046",
            },
            Object {
              "name": "Avengers by Brian Michael Bendis: The Complete Collection Vol. 2 (2017)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23600",
            },
            Object {
              "name": "Avengers Vol. 2: Red Zone (2003)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/227",
            },
            Object {
              "name": "Avengers Vol. II: Red Zone (2003)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/271",
            },
            Object {
              "name": "Captain America (1998 - 2002)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1997",
            },
            Object {
              "name": "Captain America (1968 - 1996)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1996",
            },
            Object {
              "name": "Captain America by Mark Waid, Ron Garney & Andy Kubert (2017)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23810",
            },
            Object {
              "name": "Defenders (1972 - 1986)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/3743",
            },
            Object {
              "name": "Incredible Hulks (2010 - 2011)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/8842",
            },
            Object {
              "name": "Indestructible Hulk (2012 - Present)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/16583",
            },
            Object {
              "name": "Iron Man (2012 - Present)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/16593",
            },
            Object {
              "name": "Iron Man (1968 - 1996)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/2029",
            },
            Object {
              "name": "Iron Man Epic Collection: Doom (2018)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23915",
            },
            Object {
              "name": "Marvel Adventures Super Heroes (2010 - 2012)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/9718",
            },
            Object {
              "name": "Marvel Masterworks: Captain America Vol. (2005)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1506",
            },
            Object {
              "name": "Marvel Masterworks: Captain America Vol. 1 - 2nd Edition (2003)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/189",
            },
          ],
          "returned": 20,
        },
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/6/20/52602f21f29ec",
        },
      },
    ],
    "type": "LISTING_CHARACTERS",
  },
]
`;

exports[`Testing character actions Testing of action UPDATE_CHARACTER 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "description": "",
        "id": 1011334,
        "modified": "2014-04-29T14:18:17-0400",
        "name": "3-D Man",
        "resourceURI": "http://gateway.marvel.com/v1/public/characters/1011334",
        "series": Object {
          "available": 3,
          "collectionURI": "http://gateway.marvel.com/v1/public/characters/1011334/series",
          "items": Array [
            Object {
              "name": "Avengers: The Initiative (2007 - 2010)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1945",
            },
            Object {
              "name": "Deadpool (1997 - 2002)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/2005",
            },
            Object {
              "name": "Marvel Premiere (1972 - 1981)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/2045",
            },
          ],
          "returned": 3,
        },
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
        },
      },
      Object {
        "description": "Rick Jones has been Hulk's best bud since day one, but now he's more than a friend...he's a teammate! Transformed by a Gamma energy explosion, A-Bomb's thick, armored skin is just as strong and powerful as it is blue. And when he curls into action, he uses it like a giant bowling ball of destruction! ",
        "id": 1017100,
        "modified": "2013-09-18T15:54:04-0400",
        "name": "A-Bomb (HAS)",
        "resourceURI": "http://gateway.marvel.com/v1/public/characters/1017100",
        "series": Object {
          "available": 1,
          "collectionURI": "http://gateway.marvel.com/v1/public/characters/1017100/series",
          "items": Array [
            Object {
              "name": "Hulk (2008 - 2012)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/3374",
            },
          ],
          "returned": 1,
        },
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16",
        },
      },
      Object {
        "description": "AIM is a terrorist organization bent on destroying the world.",
        "id": 1009144,
        "modified": "2013-10-17T14:41:30-0400",
        "name": "A.I.M.",
        "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009144",
        "series": Object {
          "available": 33,
          "collectionURI": "http://gateway.marvel.com/v1/public/characters/1009144/series",
          "items": Array [
            Object {
              "name": "Ant-Man & the Wasp (2010 - 2011)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/13082",
            },
            Object {
              "name": "Avengers (1998 - 2004)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/354",
            },
            Object {
              "name": "Avengers (1963 - 1996)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1991",
            },
            Object {
              "name": "Avengers and Power Pack (2017)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23123",
            },
            Object {
              "name": "Avengers and Power Pack Assemble! (2006)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1046",
            },
            Object {
              "name": "Avengers by Brian Michael Bendis: The Complete Collection Vol. 2 (2017)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23600",
            },
            Object {
              "name": "Avengers Vol. 2: Red Zone (2003)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/227",
            },
            Object {
              "name": "Avengers Vol. II: Red Zone (2003)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/271",
            },
            Object {
              "name": "Captain America (1998 - 2002)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1997",
            },
            Object {
              "name": "Captain America (1968 - 1996)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1996",
            },
            Object {
              "name": "Captain America by Mark Waid, Ron Garney & Andy Kubert (2017)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23810",
            },
            Object {
              "name": "Defenders (1972 - 1986)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/3743",
            },
            Object {
              "name": "Incredible Hulks (2010 - 2011)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/8842",
            },
            Object {
              "name": "Indestructible Hulk (2012 - Present)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/16583",
            },
            Object {
              "name": "Iron Man (2012 - Present)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/16593",
            },
            Object {
              "name": "Iron Man (1968 - 1996)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/2029",
            },
            Object {
              "name": "Iron Man Epic Collection: Doom (2018)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/23915",
            },
            Object {
              "name": "Marvel Adventures Super Heroes (2010 - 2012)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/9718",
            },
            Object {
              "name": "Marvel Masterworks: Captain America Vol. (2005)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/1506",
            },
            Object {
              "name": "Marvel Masterworks: Captain America Vol. 1 - 2nd Edition (2003)",
              "resourceURI": "http://gateway.marvel.com/v1/public/series/189",
            },
          ],
          "returned": 20,
        },
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/6/20/52602f21f29ec",
        },
      },
    ],
    "type": "LISTING_CHARACTERS",
  },
  Object {
    "payload": Object {
      "description": "",
      "id": 1011334,
      "modified": "2014-04-29T14:18:17-0400",
      "name": "Testing Update",
    },
    "type": "UPDATE_CHARACTER",
  },
]
`;
